hail/src/main/scala/is/hail/expr/ir/PValue.scala

```diff
@@ -1,27 +1,92 @@
 package is.hail.expr.ir
 
-import is.hail.asm4s.{Code, TypeInfo}
-import is.hail.expr.types.physical.{PType, PVoid}
+import is.hail.annotations.{Region, UnsafeUtils}
+import is.hail.asm4s._
+import is.hail.expr.types.physical.{PCanonicalArray, PCanonicalDict, PCanonicalSet, PContainer, PType, PVoid}
 
-abstract class PSettable {
-  def load(): PValue
+abstract class PSettable[PV <: PValue] {
+  def load(): PV
 
-  def store(v: PValue): Code[Unit]
+  def store(v: PV): Code[Unit]
 
-  def :=(v: PValue): Code[Unit] = store(v)
+  def :=(v: PV): Code[Unit] = store(v)
+
+  def storeAny(v: PValue): Code[Unit] = store(v.asInstanceOf[PV])
 }
 
 object PValue {
+  def apply(pt: PType, code: Code[_]): PValue = pt match {
+    case pt: PCanonicalArray =>
+      new PCanonicalIndexableValue(pt, coerce[Long](code))
+    case pt: PCanonicalSet =>
+      new PCanonicalIndexableValue(pt, coerce[Long](code))
+    case pt: PCanonicalDict =>
+      new PCanonicalIndexableValue(pt, coerce[Long](code))
+
+    case _ =>
+      new PrimitivePValue(pt, code)
+  }
+
   def _empty: PValue = PValue(PVoid, Code._empty)
 }
 
-case class PValue(
-  pt: PType,
-  code: Code[_]) {
+abstract class PValue {
+  def pt: PType
+
+  def code: Code[_]
+
+  def typeInfo: TypeInfo[_] = typeToTypeInfo(pt)
+
   def tcode[T](implicit ti: TypeInfo[T]): Code[T] = {
     assert(ti == typeInfo)
     code.asInstanceOf[Code[T]]
   }
+}
 
-  def typeInfo: TypeInfo[_] = typeToTypeInfo(pt)
+class PrimitivePValue(val pt: PType, val code: Code[_]) extends PValue
+
+abstract class PIndexableValue extends PValue {
+  def loadLength(): Code[Int]
+
+  def isElementDefined(i: Code[Int]): Code[Boolean]
+
+  def loadElement(length: Code[Int], i: Code[Int]): PValue
+
+  def loadElement(i: Code[Int]): PValue = loadElement(loadLength(), i)
+
+  def isElementMissing(i: Code[Int]): Code[Boolean] = !isElementDefined(i)
```

Comment:
I don't see the need for both isElementMissing and isElementDefined.

Reply:
We do it everywhere else.  I think code locally in context might prefer to be written one way or another, which is why both get there in the first place.  In particular, `defined.orEmpty` and `missing.orEmpty` are not symmetric.